core/chains/starknet/chain.go

```diff
@@ -33,7 +33,7 @@ type chain struct {
 	txm  txm.StarkTXM
 }
 
-func newChain(id string, cfg config.Config, ks keystore.StarkNet, cfgs types.Configs, lggr logger.Logger) (*chain, error) {
+func newChain(id string, cfg config.Config, ksAdapter keystore.StarkNetKeystoreAdapter, cfgs types.Configs, lggr logger.Logger) (*chain, error) {
```

Comment:
Does this need to accept the whole `StarkNetKeystoreAdapter` interface? For solana, we just have a `loop.Keystore` and that goes all the way up to the opts.

Reply:
we could do it either way. ultimately, i think it's better if it does accept the adapter because there is a strong constraint on the loop keystore -- namely that it can be used to construct an adapter and more specifically that it encodes the raw signature correctly for caigo.

i noted this in the `client.go` code, because it the constraints are not obvious and code reads a bit oddly and we wouldn't find a problem until runtime.

while writing this change, i saw that the txm in `chainlink-starknet` is taking a loop.Keystore interface and creating an adapter. that made more sense when the adapter lived in that repo.  now, i think it should take an adapter so we don't have to leave notes all over the code about these sublties.

i'll make that change after this is merged though. that's more of a touch up and maintainence.

Reply:
Sorry, I'm not following. Regardless of any particular context specific to this situation, why would we ever accept a more complex interface than is necessary? It burdens callers with creating more than they need, and e.g. complicates test code.

Reply:
i.e. we are breaking local style as well a fundamental idiom - what is the exception?

Reply:
the starknet chain doesn't work without a keystore adapter.

trying to hide that from the caller forces the caller to think even more deeply than if the adapter is required at the call site.

Reply:
Let me try to be more precise. The only usage of `ksAdapter` is to call `.Loopp()` on line 50. Why don't we just accept the interface that gets used instead of complicating things with this indirection?
```suggestion
func newChain(id string, cfg config.Config, ks loop.Keystore, cfgs types.Configs, lggr logger.Logger) (*chain, error) {
```

Reply:
you are right.

my comment above re: txm was meant to say that the current `txm.New(..., loop.Keystore,...)` doesn't make sense and that it should be txm.New(..., StarkNetKeystoreAdapter,...). 

i was proposing to change that after this.

i can use loop.Keystore in this change and then change other repo. or i can change the other repo and then make this change -- that's probably better in so much as it's fewer total changes.

in any case, the end state should be the caller provides a keystore adapter in order to create a starknet chain.

wdyt?

Reply:
I still don't understand why we would burden the caller with creating an adapter than can be created internally. What is the advantage of that change? I only see downsides.

Reply:
It seems like limiting the adapter to an internal implementation detail of the txm may also eliminate the need for the duplicate implementation here in core.

Reply:
> an adapter than can be created internally.
 
 this is the key point where we differ. a properly functioning adapter cannot be created from any loop.Keystore.
 
 it can only function correctly with a loop.Keystore implementation that does the proper encoding of the signature.

i think i've made a mistake by exposing StarkNetLoopKeystore. Instead, we should create the adapter directly, without instantiating a loop keystore, and if/when a loop keystore is needed, it can be extracted from the adapter
 
that is, the adapter is the primary interface for the chain and the loop keystore is an internal detail that is there to satisfy the relayer interface.
 
i'll code it tomorrow to give us some specific to talk about.

Reply:
I think something has become very confused. I am suggesting a trivially simplification that will allow us to remove tons of extra boilerplate.

We ultimately _cannot_ impose anything on the callers other than having  a `loop.Keystore`, because that is all we have to send over the wire. Therefore accepting any other types at this level necessarily burdens those callers with doing something to construct that different type, but we can already do that internally. Doing it internally does not introduce any additional risk that the 'wrong kind' of `loop.Keystore` will be used, because the alternative is just requiring the caller to prepackage the `loop.Keystore`  (_any_ `loop.Keystore`) in the exact same way, without any additionally accountability.

Reply:
This is what I mean: https://github.com/smartcontractkit/chainlink/pull/9397

Reply:
>Doing it internally does not introduce any additional risk that the 'wrong kind' of loop.Keystore will be used, because the alternative is just requiring the caller to prepackage the loop.Keystore (any loop.Keystore) in the exact same way, without any additional accountability.

we differ here again. you are technically correct - the relayer factory force us to pass a loop relayer at somepoint. however we can enscapulate that friction and in doing so handle error cases much more clearly

yes, it's a little more code. but the intention is much clearer and the production support is easier.

if we don't have some kind of encapsulation then the failure mode is run time error with `failed to sign` tx and a bunch jibberish big int and byte values. you would have to look at 3 different repos to try to understand that failure.

if we do encapuslate it, then we can fail much earlier with messages like `cannot create stark relayer: invalid keystore adapter"

Reply:
I would describe what I am advocating for as encapsulation. Your approach leaks internal implementation details and requires the caller to handle them, which expands the surface area for risk.

I don't follow how your approach prevents those runtime errors. The code still must ultimately accept the `loop.Keystore` interface, just at a higher level. The error may happen in a different place or sooner, but it is not fundamentally different. 

> `cannot create stark relayer: invalid keystore adapter"

You cannot know this. There are not pure Go types to check. Runtime failure is the only mode.

Reply:
The runtime error could be improved by encoding the type of signature being requested inside of the bytes payload - that is what I was advocating for with `Oneof`. The current design fundamentally leaves us with no other options than to just call it and see what happens.